### ISSUE TITLE
EPS does not notice when a connection is dead (IOError)

### ISSUE THREAD

**Issue by micressor** [chris-belcher/electrum-personal-server#259]
### Steps to reproduce:

1. Connect electrum/bitbox to your EPS via smartphone
2. Set smartphone into flight mode
3. After 20 minutes, still no new connection can be opened.

`openssl s_client -connect epsserver:50002`

The timeout of the socket came only after 25 minutes

`DEBUG: 16:16: IOError: TimeoutError(110, 'Connection timed out')`

### Workaround:

`systemctl restart electrumpersonalserver`

I think something with the connection management is not running ideally.
My wish is, that as long as EPS can only handle one connection, the timeout should be very short.

https://github.com/chris-belcher/electrum-personal-server/blob/cb6f76b0b39f059d2436fbc49e2761255e9a2559/electrumpersonalserver/server/common.py#L160

versions: eps 0.2.1.1

**Comment by chris-belcher** [comment]
Thanks for the detailed issue.

Probably the best way to solve this is to have the server ping the client, there is a protocol method for pinging: https://github.com/spesmilo/electrumx/blob/master/docs/protocol-methods.rst#serverping

There is already a place in the code to add such a ping: https://github.com/chris-belcher/electrum-personal-server/blob/6909e76167dde446d7360d278956ebe6b976e24b/electrumpersonalserver/server/common.py#L198-L207 the same place where new transactions and blocks are checked for, and the same place where the mempool is queried.

If not sending a ping, those docs say that servers can disconnect clients after a timeout.

Just so everyone knows, unfortunately this issue cant be a priority for me for the next year or so. If someone else codes it I can help review though.

**Comment by micressor** [comment]
@chris-belcher Thank you for your answer. Isn't there an easier way to set the 25 minute timeout further down? Of course, implementing the ping method is the best method in the long-term.

**Comment by chris-belcher** [comment]
You're right, there might be. The TCP socket option called keepalive might do the the trick https://www.programcreek.com/python/example/4925/socket.SO_KEEPALIVE

But if you're talking about the socket.settimeout() method which is invoked a couple of times near that code, that wont be it. That kind of timeout just means the interval at which the socket stops waiting for stuff to happen and instead sends control flow somewhere else. In EPS this feature is used to query the bitcoin core node for information about new transactions, new blocks or the state of the mempool.

**Comment by chris-belcher** [comment]
I realize that the protocol method I linked only works from the client to the server, not from the server to the client.

I tried making a hanging connection to public electrum servers out there, I just made a connection and didn't send anything. The servers don't send any kind of ping and leave the connection open for a long time (I gave up wait after about 30 minutes).

So this issue doesn't have an easy fix. The fundamental problem comes from only supporting one connection at a time, so it will be fixed when issue #50 is implemented.
